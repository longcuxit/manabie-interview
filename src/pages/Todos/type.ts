import { TodoModel, TodoStatus } from "models/Todo";
import { FocusEventHandler } from "react";

export type StatusFilter = TodoStatus | "ALL";

export interface TodoFilterState {
  keyword: string;
  status: StatusFilter;
}

export interface TodoFilterActions {
  keyword(keyword: string): void;
  status(status: StatusFilter): void;
}

export interface TodoItemProps {
  item: TodoModel;
}

export interface TodoFormProps {
  defaultValue?: string;
  onSubmit: (value: string) => Promise<void>;
  onBlur?: FocusEventHandler<HTMLFormElement>;
}
