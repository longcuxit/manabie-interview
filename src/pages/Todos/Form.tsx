import classNames from "classnames";
import { useLoading } from "components/Loader";
import { FocusEventHandler, FormEvent, MouseEvent, useState } from "react";
import { InputGroup, FormControl, Spinner } from "react-bootstrap";
import { Send, PencilFill } from "react-bootstrap-icons";

export interface TodoFormProps {
  defaultValue?: string;
  onSubmit: (value: string) => Promise<void>;
  onBlur?: FocusEventHandler<HTMLFormElement>;
}

export default function TodoForm({
  defaultValue = "",
  onSubmit,
  onBlur,
}: TodoFormProps) {
  const [value, setValue] = useState(defaultValue);
  const [loading, pushLoader] = useLoading();

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!value) return;
    await pushLoader(onSubmit(value));
    setValue("");
  };
  const handleBtnFocus = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    const input = currentTarget.previousElementSibling as HTMLInputElement;
    input.focus();
  };

  return (
    <InputGroup
      as="form"
      className={classNames(loading && "disabled")}
      onSubmit={handleSubmit}
      onBlur={onBlur}
    >
      <FormControl
        autoFocus
        value={value}
        placeholder="What you will do ..."
        onChange={({ target }) => setValue(target.value.trimLeft())}
      />
      {loading ? (
        <InputGroup.Text>
          <Spinner animation="border" size="sm" />
        </InputGroup.Text>
      ) : value ? (
        <InputGroup.Text as="button" type="submit" className="btn-send">
          <Send />
        </InputGroup.Text>
      ) : (
        <InputGroup.Text
          as="button"
          className="btn-focus"
          type="button"
          onClick={handleBtnFocus}
        >
          <PencilFill className="icon-pencil-fill" />
        </InputGroup.Text>
      )}
    </InputGroup>
  );
}
