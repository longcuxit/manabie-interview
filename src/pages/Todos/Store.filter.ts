import { TodoStatus } from "models/Todo";
import { debounce } from "utils";
import { Store } from "utils/Store";

export type StatusFilter = TodoStatus | "ALL";

interface State {
  keyword: string;
  status: StatusFilter;
}

const initialState: State = {
  keyword: "",
  status: "ALL",
};

const store = new Store(initialState, ({ partial }) => ({
  keyword: debounce((keyword: string) => partial({ keyword }), 300),
  status: (status: StatusFilter) => partial({ status }),
}));

export const useTodoFilter = store.createHook();
