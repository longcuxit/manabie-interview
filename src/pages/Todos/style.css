.Page-todos .label-big-holder {
  margin: -8px;
  padding: 8px;
  cursor: pointer;
}
