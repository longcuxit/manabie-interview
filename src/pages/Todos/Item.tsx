import { FocusEvent, useState } from "react";
import { CloseButton, FormCheck, ListGroup, Spinner } from "react-bootstrap";

import classNames from "classnames";

import { TodoStatus } from "models/Todo";
import { useTodoActions } from "store/Todos";
import { useLoading, withLoader } from "components/Loader";
import { useAsyncConfirm } from "components/AsyncModal";
import { createLocalStorage } from "utils";
import TodoForm from "./Form";
import { TodoItemProps } from "./type";

const askMeDelete = createLocalStorage("DeleteAnItem", true);

function TodoItem({ item }: TodoItemProps) {
  const actions = useTodoActions();
  const showConfirm = useAsyncConfirm();
  const [loading, pushLoader] = useLoading();

  const [editMode, setEditMode] = useState(false);

  const isActive = item.status === TodoStatus.ACTIVE;

  const handleDelete = () => {
    return pushLoader(
      (async () => {
        const isRemove =
          !askMeDelete.get() ||
          (await showConfirm({
            title: "Delete todo",
            message: "Are you sure delete this Todo?",
            footer: (
              <FormCheck
                className="flex-fill"
                label="Don't ask again!"
                onChange={({ target }) => askMeDelete.set(!target.checked)}
              />
            ),
          }));
        if (isRemove) await actions.remove(item.id);
      })()
    );
  };

  const handleBlur = (e: FocusEvent<HTMLElement>) => {
    const related = e.relatedTarget as HTMLElement;
    if (related && related.parentElement === e.currentTarget) return;
    setEditMode(false);
  };

  return (
    <ListGroup.Item
      className={classNames(
        "d-flex flex-row p-2 align-items-center",
        loading && "disabled"
      )}
      onDoubleClick={() => setEditMode(true)}
    >
      {editMode ? (
        <TodoForm
          defaultValue={item.content}
          onBlur={handleBlur}
          onSubmit={async (value: string) => {
            await actions.updateContent(item.id, value);
            setEditMode(false);
          }}
        />
      ) : (
        <>
          <label className="label-big-holder">
            <FormCheck
              checked={!isActive}
              onChange={() =>
                pushLoader(actions.updateStatus(item.id, !isActive))
              }
            />
          </label>
          <div className="flex-fill px-2">{item.content}</div>
          {loading ? (
            <Spinner animation="border" size="sm" className="mx-1" />
          ) : (
            <CloseButton onClick={handleDelete} />
          )}
        </>
      )}
    </ListGroup.Item>
  );
}
export default withLoader(TodoItem);
