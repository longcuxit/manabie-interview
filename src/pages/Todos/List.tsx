import { useMemo } from "react";
import { Container, ListGroup } from "react-bootstrap";

import { useTodos } from "store/Todos";
import TodoItem from "./Item";
import { useTodoFilter } from "./Store.filter";

export default function TodoList() {
  const [{ todos }] = useTodos();
  const [{ status, keyword }] = useTodoFilter();

  const visible = useMemo(() => {
    let visible = Array.from(todos.values());

    if (status !== "ALL") {
      visible = visible.filter((todo) => todo.status === status);
    }

    if (keyword.length) {
      const lowerKey = keyword.toLowerCase();
      visible = visible.filter((todo) =>
        todo.content.toLowerCase().includes(lowerKey)
      );
    }

    return visible;
  }, [status, keyword, todos]);

  return (
    <Container className="my-2">
      <ListGroup className="ToDo__list">
        {visible.map((todo) => (
          <TodoItem key={todo.id} item={todo} />
        ))}
      </ListGroup>
    </Container>
  );
}
