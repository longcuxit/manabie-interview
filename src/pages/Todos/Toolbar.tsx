import { MouseEvent } from "react";
import {
  Navbar,
  Container,
  DropdownButton,
  Dropdown,
  Button,
  InputGroup,
  FormControl,
  FormCheck,
} from "react-bootstrap";
import { BoxArrowRight, Search, Trash, XLg } from "react-bootstrap-icons";

import { createLocalStorage } from "utils";
import { TodoStatus } from "models/Todo";
import { useAuthAction } from "store/Auth";
import { useTodos } from "store/Todos";

import { usePushLoader } from "components/Loader";
import { PartialCheckbox } from "components/PartialCheckbox";
import { useAsyncConfirm } from "components/AsyncModal";

import { useTodoFilter } from "./Store.filter";
import { StatusFilter } from "./type";

const statusList: StatusFilter[] = [
  "ALL",
  TodoStatus.ACTIVE,
  TodoStatus.COMPLETED,
];

const askMeDeleteAll = createLocalStorage("DeleteAllItems", true);

const LogoutButton = () => {
  const authAction = useAuthAction();
  return (
    <Button className="ms-2" size="sm" onClick={authAction.logout}>
      <BoxArrowRight />
    </Button>
  );
};

export default function TodoToolbar() {
  const showConfirm = useAsyncConfirm();
  const pushLoader = usePushLoader();

  const [{ todos, countActive }, todoActions] = useTodos();
  const [{ status, keyword }, filterActions] = useTodoFilter();

  const handleDelete = async () => {
    const isRemove =
      !askMeDeleteAll.get() ||
      (await showConfirm({
        title: "Delete all completed todo",
        message: "Are you sure to clear all completed Todo?",
        footer: (
          <FormCheck
            className="flex-fill"
            label="Don't ask again!"
            onChange={({ target }) => askMeDeleteAll.set(!target.checked)}
          />
        ),
      }));
    if (isRemove) pushLoader(todoActions.clearAll());
  };

  const handleClear = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    const input = currentTarget.previousElementSibling as HTMLInputElement;
    filterActions.keyword("");
    input.value = "";
  };

  const handleFocus = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    const input = currentTarget.previousElementSibling as HTMLInputElement;
    input.focus();
  };

  return (
    <Navbar bg="primary" className="sticky-top">
      <Container>
        {Boolean(todos.size) && (
          <label className="label-big-holder ">
            <PartialCheckbox
              className="mx-2"
              checked={
                countActive === todos.size
                  ? false
                  : countActive === 0
                  ? true
                  : undefined
              }
              onChange={({ target }) =>
                pushLoader(todoActions.toggleAll(!target.checked))
              }
            />
          </label>
        )}
        <InputGroup size="sm">
          <FormControl
            placeholder="Search ..."
            onChange={({ target }) => {
              filterActions.keyword(target.value.trimLeft());
            }}
          />
          {keyword ? (
            <InputGroup.Text
              as="button"
              className="btn-clear"
              onClick={handleClear}
            >
              <XLg />
            </InputGroup.Text>
          ) : (
            <InputGroup.Text
              as="button"
              className="btn-focus"
              onClick={handleFocus}
            >
              <Search />
            </InputGroup.Text>
          )}
        </InputGroup>
        <DropdownButton
          title={status}
          onSelect={(key) => filterActions.status(key as StatusFilter)}
          size="sm"
          className="mx-2"
          align="end"
        >
          {statusList.map((item) => {
            return (
              <Dropdown.Item key={item} eventKey={item}>
                {item}
              </Dropdown.Item>
            );
          })}
        </DropdownButton>
        <Button
          size="sm"
          onClick={handleDelete}
          variant="danger"
          disabled={todos.size === countActive}
        >
          <Trash />
        </Button>
        <LogoutButton />
      </Container>
    </Navbar>
  );
}
