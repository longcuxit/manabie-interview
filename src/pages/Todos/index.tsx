import { useEffect } from "react";
import Container from "@mui/material/Container";

import { usePushLoader } from "components/Loader";
import { useTodoActions } from "store/Todos";

import TodoForm from "./Form";
import TodoList from "./List";
import TodoToolbar from "./Toolbar";

import "./style.css";

const ToDoPage = () => {
  const todoActions = useTodoActions();
  const pushLoader = usePushLoader();

  useEffect(() => {
    pushLoader(todoActions.initialize());
  }, [pushLoader, todoActions]);

  return (
    <div className="Page-todos">
      <TodoToolbar />
      <TodoList />
      <Container maxWidth="sm" sx={{ py: 1 }}>
        <TodoForm onSubmit={todoActions.create} />
      </Container>
    </div>
  );
};

export default ToDoPage;
