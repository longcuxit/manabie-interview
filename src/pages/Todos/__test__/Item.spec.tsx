import { act, fireEvent, render } from "@testing-library/react";
import { TodoStatus } from "models/Todo";
import Item from "../Item";

import { useTodoActions } from "store/Todos";
import { useAsyncConfirm } from "components/AsyncModal";
import { useLoading } from "components/Loader";
import { fakeTodo } from "utils/testing";
import TodoForm from "../Form";

jest.mock("store/Todos", () => ({ useTodoActions: jest.fn() }));
jest.mock("components/AsyncModal", () => ({ useAsyncConfirm: jest.fn() }));
jest.mock("components/Loader", () => ({
  useLoading: jest.fn(),
  withLoader: (Com: any) => Com,
}));
jest.mock("../Form", () => jest.fn());

const todo = fakeTodo({ content: "Fix content", status: TodoStatus.ACTIVE });

const mockUseTodoActions = useTodoActions as jest.Mock;
const mockUseAsyncConfirm = useAsyncConfirm as jest.Mock;
const mockUseLoading = useLoading as jest.Mock;
const mockTodoForm = TodoForm as jest.Mock;

describe("pages/Todos/Item", () => {
  const pushLoader = jest.fn();
  const updateStatus = jest.fn();
  const showConfirm = jest.fn();
  const remove = jest.fn();
  const updateContent = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    mockTodoForm.mockReturnValue(
      <div className="todo-form">
        <button />
      </div>
    );
    mockUseLoading.mockReturnValue([false, pushLoader]);
    mockUseTodoActions.mockReturnValue({ updateStatus, remove, updateContent });
    mockUseAsyncConfirm.mockReturnValue(showConfirm);
  });

  it("should match snapshot", () => {
    const { container } = render(<Item item={todo} />);

    expect(container).toMatchSnapshot();
  });

  it("should show spinner loading", () => {
    mockUseLoading.mockReturnValue([true]);

    const { container } = render(<Item item={todo} />);

    expect(container.querySelector(".spinner-border")).toBeInTheDocument();
  });

  it("should update status", () => {
    mockUseLoading.mockReturnValue([true, pushLoader]);

    const { container } = render(<Item item={todo} />);

    fireEvent.click(container.querySelector(".form-check-input")!);

    expect(pushLoader).toBeCalledTimes(1);
    expect(updateStatus).toBeCalledTimes(1);
  });

  it("should show confirm % cancel delete", async () => {
    showConfirm.mockResolvedValue(undefined);

    const { container } = render(<Item item={todo} />);

    await act(async () => {
      fireEvent.click(container.querySelector(".btn-close")!);
    });

    expect(showConfirm).toBeCalled();
    expect(pushLoader).toBeCalled();
    expect(remove).not.toBeCalled();
  });

  it("should show confirm % accept delete", async () => {
    showConfirm.mockResolvedValue(true);

    const { container } = render(<Item item={todo} />);

    await act(async () => {
      fireEvent.click(container.querySelector(".btn-close")!);
    });

    expect(showConfirm).toBeCalled();
    expect(pushLoader).toBeCalled();
    expect(remove).toBeCalled();
  });

  it("should ignore show confirm", async () => {
    const { container, rerender } = render(
      <>
        <Item item={todo} />
        {false}
      </>
    );

    await act(async () => {
      fireEvent.click(container.querySelector(".btn-close")!);
    });

    expect(showConfirm).toBeCalled();

    rerender(
      <>
        <Item item={todo} />
        {showConfirm.mock.calls[0][0].footer}
      </>
    );

    const askMe = container.querySelector(".flex-fill .form-check-input")!;

    expect(askMe).toBeInTheDocument();
    expect(localStorage.getItem("DeleteAnItem")).toBe("true");
    fireEvent.click(askMe);
    expect(localStorage.getItem("DeleteAnItem")).toBe("false");
  });

  it("should don't show confirm % accept delete", async () => {
    showConfirm.mockResolvedValue(true);

    const { container } = render(<Item item={todo} />);

    await act(async () => {
      fireEvent.click(container.querySelector(".btn-close")!);
    });

    expect(showConfirm).not.toBeCalled();
    expect(pushLoader).toBeCalled();
    expect(remove).toBeCalled();
  });

  describe("edit mode: ", () => {
    it("should open on double click", () => {
      const { container } = render(<Item item={todo} />);

      act(() => {
        fireEvent.dblClick(container.querySelector(".list-group-item")!);
      });

      expect(mockTodoForm).toBeCalled();
      expect(mockTodoForm.mock.calls[0][0].defaultValue).toBe(todo.content);
      expect(container.querySelector(".todo-form")).toBeInTheDocument();
    });
  });

  it("should edit content", async () => {
    const { container } = render(<Item item={todo} />);

    act(() => {
      fireEvent.dblClick(container.querySelector(".list-group-item")!);
    });

    expect(mockTodoForm).toBeCalled();

    await act(async () => {
      await mockTodoForm.mock.calls[0][0].onSubmit("new content");
    });

    expect(updateContent).toBeCalledWith(todo.id, "new content");
    expect(container.querySelector(".todo-form")).not.toBeInTheDocument();
  });

  it("should close on blur outside", async () => {
    const { container } = render(<Item item={todo} />);

    act(() => {
      fireEvent.dblClick(container.querySelector(".list-group-item")!);
    });

    expect(mockTodoForm).toBeCalled();

    await act(async () => {
      await mockTodoForm.mock.calls[0][0].onBlur({});
    });

    expect(container.querySelector(".todo-form")).not.toBeInTheDocument();
  });

  it("should skip close on blur inside", async () => {
    const { container } = render(<Item item={todo} />);

    act(() => {
      fireEvent.dblClick(container.querySelector(".list-group-item")!);
    });

    expect(mockTodoForm).toBeCalled();

    const form = container.querySelector(".todo-form")!;
    const button = form.querySelector("button");

    await act(async () => {
      await mockTodoForm.mock.calls[0][0].onBlur({
        currentTarget: form,
        relatedTarget: button,
      });
    });

    expect(container.querySelector(".todo-form")).toBeInTheDocument();
  });
});
