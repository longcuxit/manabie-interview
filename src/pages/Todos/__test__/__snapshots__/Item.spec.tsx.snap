// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Todos/Item should match snapshot 1`] = `
<div>
  <div
    class="d-flex flex-row p-2 align-items-center list-group-item"
  >
    <label
      class="label-big-holder"
    >
      <div
        class=""
      >
        <input
          class="form-check-input"
          type="checkbox"
        />
      </div>
    </label>
    <div
      class="flex-fill px-2"
    >
      Fix content
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div>
</div>
`;
